test/e2e/apps/job.go

```diff
@@ -465,6 +465,78 @@ done`}
+		job.Spec.MaxFailedIndexes = ptr.To[int32](0)
+
+		mode := batchv1.IndexedCompletion
+		job.Spec.CompletionMode = &mode
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Awaiting for the job to fail as the number of max failed indexes is exceeded")
+		err = e2ejob.WaitForJobFailed(f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to ensure job completion in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Verifying the Job status fields to ensure early termination of the job")
+		job, err = e2ejob.GetJob(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to retrieve latest job object")
+		gomega.Expect(job.Status.FailedIndexes).Should(gomega.HaveValue(gomega.Equal("0")))
+		gomega.Expect(job.Status.Failed).Should(gomega.Equal(int32(1)))
+	})
+
+	/*
+		Testcase: Mark indexes as failed when the FailIndex action is matched in podFailurePolicy
+		Description: Create an indexed job with backoffLimitPerIndex, and podFailurePolicy
+		with the FailIndex action. Verify the failed pods matching the pod failure policy
+		result in marking the corresponding indexes as failed without restarts, despite
+		backoffLimitPerIndex > 0.
+	*/
+	ginkgo.It("should mark indexes as failed when the FailIndex action is matched in podFailurePolicy", func(ctx context.Context) {
+		completions := int32(2)
+
+		ginkgo.By("Creating an indexed job with failing pods matching the FailIndex action")
+		job := e2ejob.NewTestJob("fail", "matching-fail-index-action", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit)
+		job.Spec.BackoffLimit = nil
+		job.Spec.BackoffLimitPerIndex = ptr.To[int32](1)
+		job.Spec.PodFailurePolicy = &batchv1.PodFailurePolicy{
+			Rules: []batchv1.PodFailurePolicyRule{
+				{
+					Action: batchv1.PodFailurePolicyActionFailIndex,
+					OnExitCodes: &batchv1.PodFailurePolicyOnExitCodesRequirement{
+						Operator: batchv1.PodFailurePolicyOnExitCodesOpIn,
+						Values:   []int32{1},
+					},
+				},
+			},
+		}
+		mode := batchv1.IndexedCompletion
+		job.Spec.CompletionMode = &mode
+		job, err := e2ejob.CreateJob(ctx, f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Awaiting for the job to fail as all indexes are failed")
+		err = e2ejob.WaitForJobFailed(f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to ensure job completion in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Verifying the Job status fields to ensure the upper indexes didn't execute")
+		job, err = e2ejob.GetJob(ctx, f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to retrieve latest job object")
+		gomega.Expect(job.Status.FailedIndexes).Should(gomega.HaveValue(gomega.Equal("0,1")))
```

Comment:
There is a guarantee on that, because the job does not specify maxFailedIndexes, so all indexes are expected to be executed.

Reply:
oh, right, it's a FailIndex, not FailJob.
It would be more interesting to use a job where (let's say, odd numbers fail). I think there's a helper for that already?

Reply:
yes, we have a helper, I was thinking about using this.